api/src/main/java/com/datatorrent/api/StatsListener.java

```diff
@@ -115,6 +117,55 @@
     List<OperatorResponse> getOperatorResponse();
   }
 
+  /**
+   * An interface to the DAG. Stats listener can get information about
+   * operator or other elements in the DAG through this interface. currerntly
+   * we only provide method to extract the operator name based on the physical
+   * id of the operator. In future more methods can be added.
+   *
+   */
+  interface StatsListenerContext
+  {
+    /**
+     * Return name of the operator given its id. Returns null if operator not found
+     * in the DAG.
+     * @param id Operator id
+     * @return name of the operator.
+     */
+    String getOperatorName(int id);
+
+    /**
+     * Create an instance of DAGChangeSet, which will be used by statsListener to submit
+     * dag modifications through {@link StatsListenerContext#submitDagChange(com.datatorrent.api.DAG.DAGChangeSet)}
+     * @return
+     */
+    DAGChangeSet createDAG();
+
+    /**
+     * Submit DAG modification request to the engine. After successful validation of
+     * new DAG, a future object is returned. StatListeners can use this future object
+     * to check the state of request. {@link FutureTask#get()} will throw an exception
+     * if any exception is thrown while DAG modifications.
+     *
+     * If an existing DAG modification is pending, then null is returned. in this case
+     * statsListener can submit the request again on next invocation.
+     *
+     * @param dagchanges The new modifications to logical dag.
+     * @return Future object to check state of the request.
+     */
+    FutureTask<Object> submitDagChange(DAGChangeSet dagchanges) throws IOException, ClassNotFoundException;
+  }
+
+  /**
+   * If StatsListener implements ContextAwareStatsListener interface, then engine will
+   * provide a reference to StatsListenerContext using which listener can examine
+   * current state of the DAG.
+   */
+  interface ContextAwareStatsListener
+  {
+    void setContext(StatsListenerContext context);
```

Comment:
Adding a new argument will break binary compatibility of existing stats listeners. hence went for this approach. This could be implemented with
- Providing context through BatchedOperatorStats.
- Adding a new interface extending StatsListener which defines processStats with additional argument, for this listeners new method will be called by platform. please suggest how can we implement this?